pwndbg/storage.py

```diff
@@ -0,0 +1,128 @@
+    try:
+        os.makedirs(dir)
+    except OSError:
+        assert os.path.isdir(dir)
+    return dir
+
+cache_dir = get_cache_dir()
+
+def makedirs(name):
+    try:
+        os.makedirs(name)
+    except OSError:  # FileExistsError.__base__ is OSError in Python 3
+        assert os.path.isdir(name)
+
+def generic_file_cache(template, fn):
+    def memoize(function):
+        p = function.__module__.split('.')[1:] + [function.__name__]
+        dirpath = os.path.join(cache_dir, *p)
+
+        @functools.wraps(function)
+        def wrapper(*args):
+            return fn(os.path.join(dirpath, template.format(*args)), function, args)
+
+        return wrapper
+
+    return memoize
+
+def file_cache_for_pickle(template):
+    def fn(path, function, args):
+        try:
+            with open(path, 'rb') as f:
+                ret = pickle.load(f)
+        except:
+            ret = function(*args)
+            makedirs(os.path.dirname(path))
+            with open(path, 'wb') as f:
+                pickle.dump(ret, f)
+
+    return generic_file_cache(template, fn)
+
+def file_cache_with_signature_for_pickle(template):
+    def fn(path, function, args):
+        valid = False
+        signature = args[-1]
+        try:
+            with open(path, 'rb') as f:
+                (saved_signature, ret) = pickle.load(f)
+                if saved_signature == signature:
+                    valid = True
+        except:
+            pass
+        if not valid:
+            ret = function(*args)
+            makedirs(os.path.dirname(path))
+            with open(path, 'wb') as f:
+                pickle.dump((signature, ret), f)
+
+    return generic_file_cache(template, fn)
+
+def file_cache_for_text(template):
+    def fn(path, function, args):
+        try:
+            with open(path) as f:
+                ret = f.read()
+        except:
+            ret = function(*args)
+            assert isinstance(ret, str)
+            makedirs(os.path.dirname(path))
+            with open(path, 'w') as f:
+                f.write(ret)
+        return ret
+
+    return generic_file_cache(template, fn)
+
+def file_cache_with_signature_for_text(template):
+    regex = re.compile(r'pwndbg:\s*signature=(\S+)')
+
+    def fn(path, function, args):
+        valid = False
+        signature = str(args[-1])
+        try:
+            with open(path) as f:
+                lines = f.readlines()
+            ret = ''
+            for i in range(len(lines)):
+                match = regex.search(lines[i])
+                if not match:
+                    ret = ''.join(lines[i:])
+                    break
+                saved_signature = match.group(1)
+                if saved_signature == signature:
+                    valid = True
+        except:
+            pass
+        if not valid:
+            ret = function(*args)
+            assert isinstance(ret, str)
```

Comment:
Probably better to have `isinstance(ret, six.string_types)` here - this would let us cache both binary/ascii strings and unicode one (however I am not sure if below file writes pass with both types).